lib/auth/webauthncli/fido2.go

```diff
@@ -0,0 +1,557 @@
+}
+
+func getPasswordlessCredentials(dev FIDODevice, pin, rpID, user string) (*libfido2.Credential, error) {
+	creds, err := dev.Credentials(rpID, pin)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	switch {
+	case len(creds) == 0:
+		return nil, libfido2.ErrNoCredentials
+	case len(creds) == 1 && user == "": // no need to disambiguate
+		cred := creds[0]
+		log.Debugf("FIDO2: Found resident credential for user %q", cred.User.Name)
+		return cred, nil
+	case len(creds) > 1 && user == "": // can't disambiguate
+		return nil, trace.BadParameter("too many credentials found, explicit user required")
+	}
+
+	duplicateWarning := false
+	var res *libfido2.Credential
+	for _, cred := range creds {
+		if cred.User.Name == user {
+			// Print information about matched credential, useful for debugging.
+			// ykman prints user IDs in hex, hence the unusual encoding choice below.
+			cID := base64.RawURLEncoding.EncodeToString(cred.ID)
+			uID := cred.User.ID
+			log.Debugf("FIDO2: Found resident credential for user %v, credential ID (b64) = %v, user ID (hex) = %x", user, cID, uID)
+			if res == nil {
+				res = cred
+				continue // Don't break, we want to warn about duplicates.
+			}
+			if !duplicateWarning {
+				duplicateWarning = true
+				log.Warnf("Found multiple credentials for %q, using first match", user)
+			}
+		}
+	}
+	if res == nil {
+		return nil, trace.BadParameter("no credentials for user %q", user)
+	}
+	return res, nil
+}
+
+func getMFACredentials(dev FIDODevice, pin, rpID, appID string, allowedCreds [][]byte) (string, []byte, error) {
+	// The actual hash is not necessary here.
+	const cdh = "00000000000000000000000000000000"
+
+	opts := &libfido2.AssertionOpts{
+		UP: libfido2.False,
+	}
+	actualRPID := rpID
+	var cID []byte
+	for _, cred := range allowedCreds {
+		_, err := dev.Assertion(rpID, []byte(cdh), [][]byte{cred}, pin, opts)
+		if err == nil {
+			cID = cred
+			break
+		}
+
+		// Try again with the U2F appID, if present.
+		if appID != "" {
+			_, err = dev.Assertion(appID, []byte(cdh), [][]byte{cred}, pin, opts)
+			if err == nil {
+				actualRPID = appID
+				cID = cred
+				break
+			}
+		}
+	}
+	if len(cID) == 0 {
+		return "", nil, libfido2.ErrNoCredentials
+	}
+
+	return actualRPID, cID, nil
+}
+
+type deviceWithInfo struct {
+	FIDODevice
+	info *deviceInfo
+}
+
+type deviceFilterFunc func(dev FIDODevice, info *deviceInfo) (ok bool, err error)
+type deviceCallbackFunc func(dev FIDODevice, info *deviceInfo, pin string) error
+
+type runPrompt interface {
+	PromptPIN() (string, error)
+	PromptAdditionalTouch() error
+}
+
+func runOnFIDO2Devices(
+	ctx context.Context,
+	prompt runPrompt, skipAdditionalPrompt bool,
+	filter deviceFilterFunc,
+	deviceCallback deviceCallbackFunc) error {
+	devices, err := findSuitableDevicesOrTimeout(ctx, filter)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	dev, requiresPIN, err := selectDevice(ctx, devices, deviceCallback)
+	if err != nil || !requiresPIN {
```

Comment:
I got a little confused on first read. Then I remembered `trace.Wrap(nil)` returns nil.

Maybe we should be explicit that `!requiresPIN` is a successful case.

```
if err != nil {
    return trace.Wrap(err)
}
if !requiresPIN {
    return nil
}
```